# STAT 440 Statistical Data Management - Fall 2021
## Challenge02
### Due: Wednesday October 13, 2021 11:59 pm US Central Time
#### Created by Christoper Kinson

Grading Rubric (per question):  
2 points if complete and correct  
1 point if incomplete or incorrect  
0 points if no attempt made  


**Retrieving your work**

This and all future challenge .md files are written in Markdown. The .md file is rendered as an .html file for easier viewing. This and all future challenge assignments are located in the **challenges** directory within the **stat440-fa21-course-content** repo, i.e. **stat440-fa21-course-content/challenges** in GHE. It is always recommended that you **pull** (or refresh) the **stat440-fa21-course-content** repo to ensure that you have the most updated version of all course content including the challenge assignments. After pulling (or refreshing) the **stat440-fa21-course-content** repo, the challenge file will be in the challenge directory. Once you have accessed the .md file, copy all text from the .md file and paste all text into your reproducible document file (either .Rmd or .ipynb file). 

**Submitting your work**

In your individual student repo (named as your netID), you are to submit ***two*** files:

a. Your reproducible document file (either .Rmd or .ipynb file) which should be saved as challenge##-netID.Rmd or challenge##-netID.ipynb depending on your preferred software. For example, my challenge 01 file would be saved as challenge01-kinson2.Rmd.

b. Your rendered reproducible document file (.html) which should be saved as challenge##-netID.html. For example, my challenge 01 file would be saved as challenge01-kinson2.html.

**Do not place your files in a sub-directory (or folder).** You have an unlimited number of submissions, but only the latest proper submission (commit and push) will be viewed and graded. You are allowed to only use one software language per challenge file. This means that you cannot mix software languages (R and Python) in one single challenge file. **Remember the .Rmd (or .ipynb) file needs to render properly to .html before submitting.** 


***

Undergraduates are expected to complete the problems **#1-#8**. Graduate students are expected to complete all problems.

**The following problems should be completed by you as an individual. If any problem asks you a particular question, be sure to answer it completely (with code, written sentences, or both). Written sentences should not appear in code chunks or code cells. Written sentences should appear in Markdown syntax unless specifically stated otherwise.**

***Do not change anything in this file above the double line. Doing so will prevent your challenge from being graded.***

***
***

### Use only one programming language for this entire challenge assignment. Use the URLs to access the data (if any). No local files allowed.

**#1** Using Markdown syntax, create a numbered list with your first name in normal text as the first item and your last name in italic text as the second item.

Answers will vary, but must be a numbered list. Graded for completeness. For example,

1. Chris

2. *Kinson*

**#2** Import the dataset(s) below using one programming language and the data URL. After importing, print the resulting last 10 rows of each dataset. Read the data key(s) and data description(s) for more information about that particular dataset.

- Speed Dating Data URL https://github-dev.cs.illinois.edu/stat440-fa21/stat440-fa21-course-content/blob/master/data/speed-dating-data.csv or Box data URL https://uofi.box.com/shared/static/w8vjq29rig8pdvussyg3997h1gwhkimq.csv  
  - Data description: the dataset (a .csv file) contains a total of 162 observations and 59 columns. The observations are four minute dates that individuals participated in for a speed dating experiment conducted by the Columbia Business School. The columns contain demographic information about the daters and datees as well as survey responses about the dater, dater's dating preferences and scores that the dater gave the datee (datee is defined as "partner"). The datee is the person of the opposite sex who joined the table for the speed date. Technically, the data should contain information from the perspective of the datee when the gender variable changes; thus making them now the dater. A match occurs when both the dater and the datee agree to go on a second date (after the experiment is over). The original sources are Ray Fisman and Sheena Iyengar.  
  - Data key URL https://github-dev.cs.illinois.edu/stat440-fa21/stat440-fa21-course-content/blob/master/data/speed-dating-data-key.doc  

In [2]:
import pandas as pd
speed_data = pd.read_csv("https://uofi.box.com/shared/static/w8vjq29rig8pdvussyg3997h1gwhkimq.csv")
speed_data.tail(10)

iid  id  gender  idg  condtn  wave  round  position  order  partner  ...  \
152  250   8       1   16       1    10      9         4      6        9  ...   
153  251   9       1   18       1    10      9         8      8        1  ...   
154  251   9       1   18       1    10      9         1      1        2  ...   
155  251   9       1   18       1    10      9         2      2        3  ...   
156  251   9       1   18       1    10      9         3      3        4  ...   
157  251   9       1   18       1    10      9         7      7        5  ...   
158  251   9       1   18       1    10      9         6      6        6  ...   
159  251   9       1   18       1    10      9         5      5        7  ...   
160  251   9       1   18       1    10      9         9      9        8  ...   
161  251   9       1   18       1    10      9         4      4        9  ...   

     attr  sinc  intel  fun  amb  shar  like  prob  met  match_es  
152     7     8    7.0  8.0  7.0   8.0   7.0   9.0  0.0         5  
153     8     8    8.0  6.0  9.0   NaN   7.0   4.0  0.0         1  
154     8     9    8.0  8.0  8.0   NaN   8.0   8.0  0.0         1  
155     4     8    7.0  3.0  8.0   NaN   3.0   6.0  0.0         1  
156     6     8    8.0  6.0  NaN   NaN   6.0   6.0  0.0         1  
157     5     8    7.0  4.0  8.0   8.0   5.0   6.0  0.0         1  
158     7     8    8.0  6.0  6.0   NaN   7.0   6.0  0.0         1  
159     7     8    7.0  5.0  8.0   NaN   5.0   6.0  0.0         1  
160     7     8    7.0  6.0  7.0   NaN   7.0   5.0  0.0         1  
161     8     9    8.0  8.0  8.0   8.0   8.0   7.0  0.0         1  

[10 rows x 59 columns]

**#3** Using the sentence about the geese below, a programming language, and string manipulation/regex to extract only the words of the sentence excluding punctuation such that each word appears as a row in a matrix.

```
#if using R
sentence <- "Friends of the Geese are hosting a memorial service Saturday for the 175 geese killed this week by the Urbana Park District in its 'charity harvest.'"

#if using Python
import pandas as pd
import numpy as np
sentence = pd.Series(["Friends of the Geese are hosting a memorial service Saturday for the 175 geese killed this week by the Urbana Park District in its 'charity harvest.'"])
```

In [8]:
import numpy as np
sentence = pd.Series(["Friends of the Geese are hosting a memorial service Saturday for the 175 geese killed this week by the Urbana Park District in its 'charity harvest.'"])
sentence01 = sentence.str.replace("\d+\s|\\.\\'", "")
sentence01.str.replace("'","").str.split(' ')[0]

C:\Users\kinson2\AppData\Local\Temp/ipykernel_11872/1689924290.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  sentence01 = sentence.str.replace("\d+\s|\\.\\'", "")


['Friends',
 'of',
 'the',
 'Geese',
 'are',
 'hosting',
 'a',
 'memorial',
 'service',
 'Saturday',
 'for',
 'the',
 'geese',
 'killed',
 'this',
 'week',
 'by',
 'the',
 'Urbana',
 'Park',
 'District',
 'in',
 'its',
 'charity',
 'harvest']

**#4** Among the matches, which iids scored the top 10 highest date_scores (an individual's date_score equals the sum of attr, sinc, intel, fun, amb, shar, like, and prob)? Use the imported data in Problem **#2**, the information in the speed dating data key, and a programming language applying data management concepts to answer the questions. **Your answer should be in Markdown syntax and a code chunk should show the result as evidence.**

The iids scoring the top 10 highest date_scores are: 238, 237, 241, 250, 251, and 244.

In [3]:
speed_data['date_score'] = speed_data.apply(lambda x: x.attr + x.sinc + x.intel + x.fun + x.amb +
                                                     x.shar + x.like + x.prob, axis=1)
speed_data02 = speed_data[(speed_data['match'] == 1)]
#success_stories = success_stories.sort_values(by='abs_int_corr', ascending=False, kind='mergesort')
speed_data03 = speed_data02.sort_values(by='date_score', ascending=False)
speed_data03.head(10)

iid  id  gender  idg  condtn  wave  round  position  order  partner  ...  \
43   238   5       0    9       1    10      9         7      9        8  ...   
34   237   4       0    7       1    10      9         3      5        8  ...   
151  250   8       1   16       1    10      9         9      2        8  ...   
70   241   8       0   15       1    10      9         9      2        8  ...   
37   238   5       0    9       1    10      9         7      3        2  ...   
148  250   8       1   16       1    10      9         7      9        5  ...   
161  251   9       1   18       1    10      9         4      4        9  ...   
152  250   8       1   16       1    10      9         4      6        9  ...   
150  250   8       1   16       1    10      9         5      7        7  ...   
94   244   2       1    4       1    10      9         7      3        5  ...   

     sinc  intel  fun   amb  shar  like  prob  met  match_es  date_score  
43      8    9.0  9.0  10.0  10.0   9.0   8.0  0.0         2        72.0  
34      8   10.0  8.0  10.0   9.0   8.0   9.0  NaN         0        68.0  
151    10    8.0  8.0   8.0   8.0   8.0   9.0  0.0         5        67.0  
70      6   10.0  9.0  10.0   9.0   8.0   7.0  0.0         2        67.0  
37      9    9.0  8.0   9.0   8.0   8.0   7.0  0.0         2        66.0  
148     8    8.0  9.0   7.0   8.0   8.0   9.0  0.0         5        65.0  
161     9    8.0  8.0   8.0   8.0   8.0   7.0  0.0         1        64.0  
152     8    7.0  8.0   7.0   8.0   7.0   9.0  0.0         5        61.0  
150     7    7.0  8.0   9.0   7.0   7.0  10.0  0.0         5        61.0  
94      8    8.0  8.0   9.0   7.0   8.0   5.0  0.0         1        61.0  

[10 rows x 60 columns]

**#5** How many interracial (dater and datee race is not the same) matches occur? Use the imported data in Problem **#2**, the information in the speed dating data key, and a programming language applying data management concepts to answer the questions. **Your answer should be in Markdown syntax and a code chunk should show the result as evidence.**

In [ ]:
There are 24 interracial matches occurring in the data.

In [4]:
speed_data.loc[(speed_data['match']==1) & (speed_data['samerace'] == 0), 'iid']

14     235
34     237
43     238
61     240
63     241
68     241
70     241
72     242
75     242
77     242
79     242
80     242
88     243
89     243
116    246
127    248
133    248
134    248
147    250
148    250
150    250
151    250
152    250
161    251
Name: iid, dtype: int64